clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java

```diff
@@ -193,6 +193,20 @@ else if (maxTimestamp == RecordBatch.NO_TIMESTAMP)
             return new RecordsInfo(maxTimestamp, compressionType == CompressionType.NONE ? offsetOfMaxTimestamp : lastOffset);
     }
 
+    public void setProducerState(long pid, short epoch, int baseSequence) {
+        if (isClosed()) {
+            // Sequence numbers are assigned when the batch is closed while the accumulator is being drained.
+            // If the resulting ProduceRequest to the partition leader failed for a retriable error, the batch will
+            // be re queued. In this case, we should not attempt to set the state again, since changing the pid and sequence
+            // once a batch has been sent to the broker risks introducing duplicates.
+            throw new IllegalStateException("Attempting to close the same batch with a different base sequence. "
+                    + "Being in this situation indicates data corruption is afoot.");
```

Comment:
Actually, the whole error message is out dated. We should not be calling `setProducerState` on a closed batch any more. Doing so is a bug on the client. Updated the message to indicate that.